### ISSUE TITLE
Timeline Explorer

### ISSUE THREAD

**Issue by vikas891** [EricZimmerman/Issues#175]
Only in the recent times MFTECmd parsed CSVs do not load up correctly in Timeline Explorer (both latest versions I regularly fire up the PS1)

```
2021-10-17 15:22:52.0352 | INFO | Timeline Explorer v1.3.0.0 ready!
2021-10-17 15:23:20.4671 | INFO | Loading 'D:\3DCaseData\89 - [Truncated]\IREC\20211016234026-[TRUNCATED]\Content\C\20211017095300_MFTECmd_$MFT_Output.csv'...
2021-10-17 15:23:21.4356 | ERROR | Error processing 'D:\3DCaseData\89 - [Truncated]\IREC\20211016234026-[TRUNCATED]\Content\C\20211017095300_MFTECmd_$MFT_Output.csv': An unexpected error occurred.
IReader state:
   ColumnCount: 0
   CurrentIndex: 19
   HeaderRecord:
["EntryNumber","SequenceNumber","InUse","ParentEntryNumber","ParentSequenceNumber","ParentPath","FileName","Extension","FileSize","ReferenceCount","ReparseTarget","IsDirectory","HasAds","IsAds","SI<FN","uSecZeros","Copied","SiFlags","NameType","Created0x10","Created0x30","LastModified0x10","LastModified0x30","LastRecordChange0x10","LastRecordChange0x30","LastAccess0x10","LastAccess0x30","UpdateSequenceNumber","LogfileSequenceNumber","SecurityId","ObjectIdFileDroid","LoggedUtilStream","ZoneIdContents"]
IParser state:
   ByteCount: 0
   CharCount: 674
   Row: 2
   RawRow: 2
   Count: 33
   RawRecord:
0,1,True,5,5,.,$MFT,,444596224,1,,False,False,False,False,False,False,Hidden|System,DosWindows,2018-08-09 21.46.56.4004254,,2018-08-09 21.46.56.4004254,,2018-08-09 21.46.56.4004254,,2018-08-09 21.46.56.4004254,,0,158679767204,256,,,

 (String was not recognized as a valid DateTime.)
2021-10-17 15:23:21.4356 | INFO | For more information, enable debug messages via View | Show Debug messages. Each successfully processed row is shown in the debug output (plugin dependent), so the bad data is generally the row following the last row shown in debug output.
2021-10-17 15:23:21.4786 | INFO | Found plugin with description 'CSV generated from MFTECmd for $MFT'
```
I mean, if Timeline Explorer fetched from any other source, I'd say its OK but I would like both tools to be on friendly terms here.

**Comment by AndrewRathbun** [comment]
I just did a test with the latest binaries and TLE opened my 600+ MB CSV without error. Can you try reloading that CSV with debug messages enabled and see if any more information is provided by TLE?

**Comment by AndrewRathbun** [comment]
Also, what is your workflow? Using MFTECmd by itself and ingesting the output into TLE? Or running KAPE and ingesting the MFTECmd output from there? If KAPE, are you sure your binary within `.\KAPE\Modules\bin` is updated? `Get-ZimmermanTools.ps1` doesn't touch that folder and `Get-KAPEUpdate.ps1` doesn't touch it either. I won't be able to answer your question since I'm not Eric but I'm just asking questions that relate to very common scenarios I see.

**Comment by vikas891** [comment]
Hi Andrew, thank you for so much for taking the time to respond. Any/all help or suggestions are always welcome! Um, no KAPE. 

Simple commands to output into CSV and then TLE

```
C:\1Tools\MFTECmd.exe -f ./$MFT --csv ./
C:\1Tools\MFTECmd.exe -f ./$Extend\$UsnJrnl\$J --csv ./
```

As we work with confidential data, it's irritating that we can't share a copy but the verbose logs on one of the CSVs give a similar error:

`2021-10-17 17:26:29.7505 | ERROR | Error processing 'D:\3DCaseData\89 - xxxxxx\IREC\XXXXXX\XXXXX_20211017105450_MFTECmd_$MFT_Output.csv': An unexpected error occurred. (String was not recognized as a valid DateTime.)`

It's practically stalled my work :-/

**Comment by AndrewRathbun** [comment]
IREC is Binalyze IREC, right?

**Comment by vikas891** [comment]
Oh right, that could be a factor. The way $MFT being collected - nice catch Andrew. And yes, TACTICAL(formerly IREC) is Binalyze :) I should perform some tests with that tool and share my observations.

**Comment by AndrewRathbun** [comment]
Yes, if the MFT was collected with Binalyze IREC (cool tool, btw, in my limited testing), I see as of 9/17/2021 they have `Improved MFT Collector` on their Release Notes, so maybe the issue is with how the MFT is collected? Source: https://binalyze.com/tactical/release-notes/

Have you tried collecting the MFT with KAPE using the appropriate Target? A workflow could be $MFT Target or FileSystem Compound Target and then parsing with either the MFTECmd_$MFT Module or MFTECmd Compound Module. Either way you'll get your CSVs for the MFT.

**Comment by EricZimmerman** [comment]
Look at the error closely

It's saying that you have a bad date or time somewhere which can either mean its malformed or it's missing out right and it's expected to be there

If you enable debug you should be able to see the exact row of data in the log file that's causing you to have a problem

This in turn can be compared to a known good row and you'll be able to find where your problem is

**Comment by vikas891** [comment]
I did! It is not telling me which row is problematic :-/ 

```
2021-10-17 18:29:31.1182 | INFO | Loading 'C:\Temp\TRASHPANDA\C\20211017125606_MFTECmd_$MFT_Output.csv'...
2021-10-17 18:29:31.2592 | ERROR | Error processing 'C:\Temp\TRASHPANDA\C\20211017125606_MFTECmd_$MFT_Output.csv': An unexpected error occurred. (String was not recognized as a valid DateTime.)
2021-10-17 18:29:31.2592 | INFO | For more information, enable debug messages via View | Show Debug messages. Each successfully processed row is shown in the debug output (plugin dependent), so the bad data is generally the row following the last row shown in debug output.
```
I tested with CyLR + TACTICAL on my test machine but TLE (or MFTECmd) don't want to play along :(

**Comment by EricZimmerman** [comment]
Perhaps it's the first row of data and it's expecting timestamps to exist and they don't

There's nothing confidential about the first 12 rows in an mft because they're always the same

**Comment by EricZimmerman** [comment]
Did you by chance override the date time format on the command line?

I don't think that you enable debug because if you look at the third row of the message you posted it says to enable debug mode so you need to start timeline Explorer enable debug mode and then reload the file

**Comment by EricZimmerman** [comment]
You can also give kape world internally and just not use it for production if it makes it easier to serve as a point of reference

**Comment by vikas891** [comment]
Weird, as a general practice I try to keep it enabled. Anyway
```
2021-10-17 18:34:28.7570 | INFO | Timeline Explorer v1.3.0.0 ready!
2021-10-17 18:34:32.9686 | DEBUG | Enabled debug messages...
2021-10-17 18:34:38.9656 | INFO | Loading 'C:\Temp\TRASHPANDA\C\20211017125606_MFTECmd_$MFT_Output.csv'...
2021-10-17 18:34:39.8059 | ERROR | Error processing 'C:\Temp\TRASHPANDA\C\20211017125606_MFTECmd_$MFT_Output.csv': An unexpected error occurred. (String was not recognized as a valid DateTime.)
2021-10-17 18:34:39.8059 | INFO | For more information, enable debug messages via View | Show Debug messages. Each successfully processed row is shown in the debug output (plugin dependent), so the bad data is generally the row following the last row shown in debug output.
```
I opened in Excel and first 12 rows did have time-stamps (or maybe I'm wrong)

Would really appreciate some help. I uploaded a CyLR + $MFT + CSV 

`https://mega.nz/folder/qsJD3ILR#qrEaKqSpq9V5KRVfIQZ5uw `

**Comment by vikas891** [comment]
Sorry, I missed that question - no I usually don't. 

`PS C:\Temp\TRASHPANDA\C> MFTECmd.exe -f '.\$MFT' --csv ./`

**Comment by EricZimmerman** [comment]
I won't be able to look at this until later this evening

**Comment by EricZimmerman** [comment]
it loaded just fine on my box.

i wonder, what are your regional settings? perhaps force the dateformat to yyyy-mm-dd HH:mm:ss.ffffff on the command line?

here are the first few rows of the file **i** generated just this morning:

```
EntryNumber,SequenceNumber,InUse,ParentEntryNumber,ParentSequenceNumber,ParentPath,FileName,Extension,FileSize,ReferenceCount,ReparseTarget,IsDirectory,HasAds,IsAds,SI<FN,uSecZeros,Copied,SiFlags,NameType,Created0x10,Created0x30,LastModified0x10,LastModified0x30,LastRecordChange0x10,LastRecordChange0x30,LastAccess0x10,LastAccess0x30,UpdateSequenceNumber,LogfileSequenceNumber,SecurityId,ObjectIdFileDroid,LoggedUtilStream,ZoneIdContents
0,1,True,5,5,.,$MFT,,549191680,1,,False,False,False,False,False,False,Hidden|System,DosWindows,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,0,7283984368,256,,,
1,1,True,5,5,.,$MFTMirr,,4096,1,,False,False,False,False,False,False,Hidden|System,DosWindows,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,0,33560283,256,,,
2,2,True,5,5,.,$LogFile,,67108864,1,,False,False,False,False,False,False,Hidden|System,DosWindows,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,0,33560353,256,,,
3,3,True,5,5,.,$Volume,,0,1,,False,False,False,False,False,False,Hidden|System,DosWindows,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,0,218860291,257,732671fc-bda4-4ec0-98ec-2a756becb8a2,,
4,4,True,5,5,.,$AttrDef,,2560,1,,False,False,False,False,False,False,Hidden|System,DosWindows,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,2021-06-03 01:19:37.0628010,,0,33560493,256,,,
5,5,True,5,5,.,.,,0,1,,True,False,False,True,False,False,Hidden|System,DosWindows,2020-11-18 23:24:04.5272620,2021-06-03 01:19:37.0628010,2021-10-17 12:33:22.8386112,2021-06-03 01:19:37.0628010,2021-10-17 12:33:22.8386112,2021-06-03 01:19:37.0628010,2021-10-17 12:53:58.3937462,2021-06-03 01:19:37.0628010,24328568,7505519304,264,ebee8895-c3ac-11eb-b69d-000c2932a55a,$TXF_DATA,
```

loaded fine in TLE

![image](https://user-images.githubusercontent.com/4265254/137748339-188e16d5-cd90-4c13-81f5-e35ad44b440d.png)

both files are exactly the same size in bytes, BUT here is the problem:

![image](https://user-images.githubusercontent.com/4265254/137750758-f0913fbe-32ab-48b9-92cf-67523927364a.png)

your data is on the bottom

TLE is expecting :. but you are using . for the time separator

also, with debug turned on, here you go

```
2021-10-18 10:12:12.8083 | INFO | Timeline Explorer v1.3.0.0 ready!
2021-10-18 10:12:20.0338 | INFO | Loading 'C:\Temp\20211018140818_MFTECmd_$MFT_Output.csv'...
2021-10-18 10:12:28.1130 | INFO | Found plugin with description 'CSV generated from MFTECmd for $MFT'
2021-10-18 10:28:32.8756 | INFO | Loading 'C:\Temp\His_20211017125606_MFTECmd_$MFT_Output.csv'...
2021-10-18 10:28:32.9586 | ERROR | Error processing 'C:\Temp\His_20211017125606_MFTECmd_$MFT_Output.csv': An unexpected error occurred.
IReader state:
   ColumnCount: 0
   CurrentIndex: 19
   HeaderRecord:
["EntryNumber","SequenceNumber","InUse","ParentEntryNumber","ParentSequenceNumber","ParentPath","FileName","Extension","FileSize","ReferenceCount","ReparseTarget","IsDirectory","HasAds","IsAds","SI<FN","uSecZeros","Copied","SiFlags","NameType","Created0x10","Created0x30","LastModified0x10","LastModified0x30","LastRecordChange0x10","LastRecordChange0x30","LastAccess0x10","LastAccess0x30","UpdateSequenceNumber","LogfileSequenceNumber","SecurityId","ObjectIdFileDroid","LoggedUtilStream","ZoneIdContents"]
IParser state:
   ByteCount: 0
   CharCount: 672
   Row: 2
   RawRow: 2
   Count: 33
   RawRecord:
0,1,True,5,5,.,$MFT,,549191680,1,,False,False,False,False,False,False,Hidden|System,DosWindows,2021-06-03 01.19.37.0628010,,2021-06-03 01.19.37.0628010,,2021-06-03 01.19.37.0628010,,2021-06-03 01.19.37.0628010,,0,7283984368,256,,,

 (String was not recognized as a valid DateTime.)
2021-10-18 10:28:32.9586 | INFO | For more information, enable debug messages via View | Show Debug messages. Each successfully processed row is shown in the debug output (plugin dependent), so the bad data is generally the row following the last row shown in debug output.
2021-10-18 10:28:32.9766 | INFO | Found plugin with description 'CSV generated from MFTECmd for $MFT'

```

look at the error message. its complaining about CurrentIndex 19, which is the Created0x10 field (starting at 0)

TL;DR;

you are using an incorrect (according to what TLE is expecting) time separator. change it on your computer to use : or override the format on the command line i suppose.
